src/reader.cpp

```diff
@@ -1626,30 +1626,7 @@ static void reader_interactive_init() {
         signal_set_handlers();
     }
 
-    // Put ourselves in our own process group.
-    shell_pgid = getpid();
-    if (getpgrp() != shell_pgid && setpgid(shell_pgid, shell_pgid) < 0) {
-        debug(0, _(L"Couldn't put the shell in its own process group"));
-        wperror(L"setpgid");
-        exit_without_destructors(1);
-    }
-
-    // Grab control of the terminal.
-    if (tcsetpgrp(STDIN_FILENO, shell_pgid) == -1) {
-        if (errno == ENOTTY) redirect_tty_output();
-        debug(0, _(L"Couldn't grab control of terminal"));
-        wperror(L"tcsetpgrp");
-        exit_without_destructors(1);
-    }
-
     invalidate_termsize();
-
-    // Set the new modes.
-    if (tcsetattr(0, TCSANOW, &shell_modes) == -1) {
```

Comment:
Unfortunately both `tcsetattr` and `tcsetpgrp` are responsible for deadlocking fish/the reader since it seems that just like fish sometimes spawns fish and waits for the process to be in a certain state before giving it control of the shell, other shells (`bash -c`) do too.

The man page for `tcgetattr` specifies it can be used from background processes, but the docs on `tcsetattr` are very skimpy and don't mention anything at all. It _seems_ from our experience that this isn't safe to call until the shell has full control of the terminal.

In the interest of laying out all options, it might be possible to use SIGALRM to force an interrupt in both these calls if they haven't responded in xx seconds, but I would rather just call them once we know we have control of the terminal at some later point.